### PR TITLE
fix(util): strip Gemini-rejected JSON Schema keywords

### PR DESCRIPTION
**Pull request by rome-xi**
## Summary

Private Gemini/Antigravity backends reject several JSON Schema keywords. Extend the existing cleaner so unsupported constraints do not reach those backends while retaining their meaning where possible:

- move `minProperties` / `maxProperties` and other unsupported constraints to description hints before removal
- preserve or infer boolean type for boolean `enum` / `const`, including nullable unions, while retaining allowed values as hints
- project integer `exclusiveMinimum` / `exclusiveMaximum` exactly with bounded arbitrary-precision arithmetic; retain exact hints for continuous or unrepresentable cases
- reject impractically large numeric literals/exponents before arbitrary-precision parsing
- recursively intersect `allOf` bounds, sizes, required fields, typed enums, and `additionalProperties` before lossy conversion; preserve non-orderable constraint hints
- keep composite `enum`, `const`, `default`, and `examples` values opaque to schema walkers so literal objects are never rewritten; preserve root const paths and large-integer spellings exactly
- distinguish author-selected names in `properties`, definition maps, pattern/dependent schema maps, and legacy hybrid `dependencies`, including schema-valued and array-valued dependencies

Property names that collide with schema keywords remain schemas and are still cleaned recursively.

## Validation

- `go test ./internal/util -count=1`
- targeted regressions repeated 10 times
- `go test -race ./internal/util -count=1`
- `go vet ./internal/util`
- server build and full diff check

Fork PR: https://github.com/rome-xi/CLIProxyAPI/pull/7

### CODE REVIEW

File: internal/util/gemini_schema.go [hunk 1]
```diff
@@ -763,6 +766,66 @@ func convertRefsToHints(jsonStr string, preserveSiblings bool) string {
...
+	}
+	return jsonStr
+}
+
+func projectIntegerExclusiveBound(raw string, exclusiveMaximum bool) string {
+	if i, err := strconv.ParseInt(strings.TrimSpace(raw), 10, 64); err == nil {
+		if exclusiveMaximum {
+			return strconv.FormatInt(i-1, 10)
+		}
+		return strconv.FormatInt(i+1, 10)
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: 已改为 `math/big` 任意精度运算，覆盖 int64 两端及更大整数，不再发生 ±1 溢出。

File: internal/util/gemini_schema.go [hunk 2]
```diff
@@ -830,7 +883,7 @@ func addEnumHints(jsonStr string) string {
 func dropIgnoredEnumsToHints(jsonStr string, options jsonSchemaCleanOptions) string {
 	for _, path := range findPaths(jsonStr, "enum") {
 		parentPath := trimSuffix(path, ".enum")
-		shouldDrop := options.dropAllEnums || (options.dropBooleanEnums && gjson.Get(jsonStr, joinPath(parentPath, "type")).String() == "boolean")
+		shouldDrop := options.dropAllEnums || (options.dropBooleanEnums && isBooleanEnumSchema(jsonStr, parentPath, gjson.Get(jsonStr, path)))
 		if !shouldDrop {
 			continue
 		}
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: 已修复：在删除不支持的 boolean enum/const 前，为无显式 type 的纯布尔约束补 `type: boolean`；Gemini、Antigravity tool/response 均有回归覆盖。

File: internal/util/gemini_schema.go [hunk 3]
```diff
@@ -763,6 +765,39 @@ func convertRefsToHints(jsonStr string, preserveSiblings bool) string {
...
+				continue
+			}
+			if val.Type == gjson.Number {
+				incPath := joinPath(parentPath, inclusive)
+				if !gjson.Get(jsonStr, incPath).Exists() {
+					updated, _ := sjson.SetRawBytes([]byte(jsonStr), incPath, []byte(val.Raw))
+					jsonStr = string(updated)
+				}
+			}
+			jsonStr, _ = sjson.Delete(jsonStr, p)
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: 已修复：integer exclusive bound 精确投影为 floor+1 / ceil-1，并与既有 inclusive sibling 选择更严格者；连续 number 保留最近可表达约束并在 description 记录 exact exclusivity。

File: internal/util/gemini_schema.go [hunk 4]
```diff
@@ -763,6 +766,95 @@ func convertRefsToHints(jsonStr string, preserveSiblings bool) string {
...
+			if isPropertyDefinition(parentPath) {
+				continue
+			}
+			val := gjson.Get(jsonStr, p)
+			if !val.Exists() {
+				continue
+			}
+
+			incPath := joinPath(parentPath, inclusive)
+			integerSchema := gjson.Get(jsonStr, joinPath(parentPath, "type")).String() == "integer"
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: 已在 6a98522a 修复：projection 现在按 flattenTypeArrays 的同一规则读取首个非 null 类型，因此 `[integer,null]` 与 `[null,integer]` 都精确投影为 integer 边界；新增 minimum=1 / maximum=9 回归测试。

File: internal/util/gemini_schema.go [hunk 5]
```diff
@@ -796,14 +906,28 @@ func convertEnumValuesToStrings(jsonStr string, forceStringType bool) string {
...
+func isBooleanEnumSchema(jsonStr, parentPath string, arr gjson.Result) bool {
+	if gjson.Get(jsonStr, joinPath(parentPath, "type")).String() == "boolean" {
+		return true
+	}
+	if !arr.IsArray() || len(arr.Array()) == 0 {
+		return false
+	}
+	for _, item := range arr.Array() {
+		if item.Type != gjson.True && item.Type != gjson.False {
+			return false
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:1042]: 已修复：显式 nullable boolean type-array 会按 effective boolean 识别，null 成员不再触发字符串化；无 type 的 bool+null 也会推断 `[boolean,null]`，并保留 nullable/Allowed hints。

File: internal/util/gemini_schema.go [hunk 6]
```diff
@@ -103,6 +105,7 @@ func cleanJSONSchema(jsonStr string, options jsonSchemaCleanOptions) string {
 		jsonStr = inlineLocalRefs(jsonStr)
 	}
 	jsonStr = convertRefsToHints(jsonStr, options.antigravitySemantics)
+	jsonStr = projectExclusiveBounds(jsonStr)
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:108]: 已修复：allOf 使用专属递归求交，minimum 取最大、maximum 取最小，含嵌套属性与跨分支 integer 类型继承；description 也合并以保留 continuous exclusive 两侧语义。

File: internal/util/gemini_schema.go [hunk 7]
```diff
@@ -763,6 +766,109 @@ func convertRefsToHints(jsonStr string, preserveSiblings bool) string {
...
+		if value := item.String(); value != "" && value != "null" {
+			return value
+		}
+	}
+	return ""
+}
+
+func projectIntegerExclusiveBound(raw string, exclusiveMaximum bool) (string, bool) {
+	value, ok := new(big.Rat).SetString(strings.TrimSpace(raw))
+	if !ok {
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:943]: 已在 774b448f 修复：所有 big.Rat 入口统一经过 bounded parser，限制字面量长度和指数绝对值；`1e±1000000000` 安全降级为 hint，existing-bound 比较也不会绕过限制。

File: internal/util/gemini_schema.go [hunk 8]
```diff
@@ -796,14 +993,51 @@ func convertEnumValuesToStrings(jsonStr string, forceStringType bool) string {
...
 			jsonStr = string(updated)
 		}
 	}
 	return jsonStr
 }
 
+func isBooleanEnumSchema(jsonStr, parentPath string, arr gjson.Result) bool {
+	schemaType := gjson.Get(jsonStr, joinPath(parentPath, "type"))
+	if schemaType.Exists() {
+		return effectiveSchemaType(schemaType) == "boolean"
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: Fixed in eadb10c1. Boolean classification now validates every enum member before consulting an explicit boolean or boolean-first union type: only boolean values (plus null for nullable schemas) take the boolean-enum path. A regression test covers type [boolean,string] with string enum values and verifies Gemini selects string and retains the enum; the previous response boolean test now uses an actual boolean literal. Full util tests, race tests, vet, server build, and diff check pass.

[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: Follow-up cb96a1ed also covers the adjacent edge cases found during audit: explicit boolean empty enums remain boolean and are dropped, untyped empty enums do not infer boolean, and boolean enums select boolean even when it is not the first member of a type union. Nullable is retained only when both the enum and declared type permit null. Regression coverage runs across Gemini and both Antigravity cleaners.

File: internal/util/gemini_schema.go [hunk 9]
```diff
@@ -763,6 +768,194 @@ func convertRefsToHints(jsonStr string, preserveSiblings bool) string {
...
+}
+
+func setStricterNumericBound(jsonStr, path, projected string, maximum bool) string {
+	existing := gjson.Get(jsonStr, path)
+	if existing.Exists() {
+		projectedValue, projectedOK := parseSchemaNumericBound(projected)
+		existingValue, existingOK := parseSchemaNumericBound(existing.Raw)
+		if !projectedOK || !existingOK {
+			return jsonStr
+		}
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:976]: Fixed in 38e3a29f. setStricterNumericBound now reports whether the incoming projected bound is representable; when an existing numeric sibling exceeds the bounded parser, the cleaner retains it and records the exact exclusive constraint as a description hint instead of silently deleting that semantic. Invalid non-numeric siblings are replaced by the valid projected bound. Lower/upper, integer/continuous, capped, and invalid sibling regressions cover all cleaners.

File: internal/util/gemini_schema.go [hunk 10]
```diff
@@ -987,6 +1246,43 @@ func mergeAllOf(jsonStr string) string {
...
+// mergeAllOfSchemaAtPath preserves the parent-first behavior for schema shape while intersecting
+// numeric bounds. Bounds may be nested under properties, so this rule is intentionally recursive
+// and local to allOf rather than changing the merger also used for lossy anyOf projections.
+func mergeAllOfSchemaAtPath(jsonStr, destination, field string, incoming gjson.Result) string {
+	if incoming.Type == gjson.Number {
+		switch field {
+		case "minimum":
+			return setStricterNumericBound(jsonStr, destination, incoming.Raw, false)
+		case "maximum":
+			return setStricterNumericBound(jsonStr, destination, incoming.Raw, true)
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go]: Fixed in 38e3a29f. allOf now intersects minProperties/minItems/minLength by maximum and the corresponding max constraints by minimum, recursively and independent of branch order. The same audit also preserves non-orderable constraint hints, unions nested required fields, intersects enums in O(n+m) with typed keys, and retains the strictest additionalProperties value. Full util/race/vet/build/diff validation passes.

File: internal/util/gemini_schema.go [hunk 11]
```diff
@@ -796,14 +1007,65 @@ func convertEnumValuesToStrings(jsonStr string, forceStringType bool) string {
...
+		switch item.Type {
+		case gjson.True, gjson.False:
+			hasBoolean = true
+		case gjson.Null:
+			// An untyped boolean enum can express nullability through a null member.
+		default:
+			return false
+		}
+	}
+	schemaType := gjson.Get(jsonStr, joinPath(parentPath, "type"))
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:1045]: Fixed in 370d2b3a. The cleaner now merges conditionals/allOf immediately after const-to-enum conversion and before enum stringification, hints, or boolean inference. A sibling type [boolean,null] is therefore visible when classifying const/enum null regardless of branch order or nesting. Regressions cover all three cleaners, both branch orders, nested schemas, and idempotence.

File: internal/util/gemini_schema.go [hunk 12]
```diff
@@ -987,6 +1263,153 @@ func mergeAllOf(jsonStr string) string {
...
+			if name := required.String(); !contains(current, name) {
+				current = append(current, name)
+			}
+		}
+		updated, _ := sjson.SetBytes([]byte(jsonStr), destination, current)
+		return string(updated)
+	}
+	if field == "enum" && incoming.IsArray() && !propertyDefinition {
+		if existing := gjson.Get(jsonStr, destination); existing.IsArray() {
+			updated, _ := sjson.SetRawBytes([]byte(jsonStr), destination, []byte(intersectEnumValues(existing, incoming)))
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:1297]: Fixed in 370d2b3a. allOf enum intersection now runs while members retain their original JSON types, so mathematically equivalent numeric spellings such as 1 and 1.0 share the same bounded canonical key before upstream string conversion. Typed intersection remains O(n+m); composite keys also canonicalize object key order while preserving array order and nested numeric equivalence. Targeted, full util, race, vet, build, and diff checks pass.

File: internal/util/gemini_schema.go [hunk 13]
```diff
@@ -853,8 +1129,8 @@ func addAdditionalPropertiesHints(jsonStr string) string {
 }
 
 var unsupportedConstraints = []string{
-	"minLength", "maxLength", "exclusiveMinimum", "exclusiveMaximum",
-	"pattern", "minItems", "maxItems", "uniqueItems", "contains", "format",
+	"minLength", "maxLength",
+	"pattern", "minItems", "maxItems", "minProperties", "maxProperties", "uniqueItems", "contains", "format",
```
[the remark this answers is not in the document]
Comment (rome-xi) [reply internal/util/gemini_schema.go:1143]: Fixed in 68734142, with the adjacent legacy dependencies name-map boundary completed in c784fa84. All recursive keyword, ref, and x-extension walkers now stop at opaque instance/annotation values such as enum, const, default, and examples, while properties/$defs/patternProperties/dependentSchemas/dependencies author-selected names remain traversable schemas. const-to-enum also preserves raw composite and large-integer values. Regressions cover nested literal objects, schema-key collisions, root const, hybrid dependencies, all three cleaners, and idempotence; full util/race/vet/build/diff validation passes.
